src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3055,13 +3053,18 @@ size_t os::pd_pretouch_memory(void* first, void* last, size_t page_size) {
     if (!UseMadvPopulateWrite || err == EINVAL) { // Not to use or not supported
       // When using THP we need to always pre-touch using small pages as the
       // OS will initially always use small pages.
-      return os::vm_page_size();
-    } else if (err != 0) {
-      log_info(gc, os)("::madvise(" PTR_FORMAT ", " SIZE_FORMAT ", %d) failed; "
-                       "error='%s' (errno=%d)", p2i(first), len,
-                       MADV_POPULATE_WRITE, os::strerror(err), err);
+      page_size = os::vm_page_size();
+    } else if (err == 0) {
+      page_size = 0;
+    }
+    if (UseMadvPopulateWrite) {
+      log_debug(gc, os)("Called madvise(" PTR_FORMAT ", " SIZE_FORMAT ", %d):"
+                        " error='%s' (errno=%d), when THPMode::always=%d and"
+                        " UseTransparentHugePages=%d",
+                        p2i(first), len, MADV_POPULATE_WRITE, os::strerror(err),
+                        err, (int)(HugePages::thp_mode() == THPMode::always),
+                        (int)UseTransparentHugePages);
```

Comment:
This function has become quite messy: It checks UseMadvPopulateWrite tree times. It mutates the input variable. It logs even when no failure occurred (This might be a bug).

I'd much prefer if we could straighten out the code into something like this (I've not tried to compile this):
```
size_t os::pd_pretouch_memory(void* first, void* last, size_t page_size) {
  if (HugePages::thp_mode() != THPMode::always && !UseTransparentHugePages) {
    // No THP. Use the platform-independent pretouch memory code.
    return page_size;
  }

  if (!UseMadvPopulateWrite) {
    // Use small pages with the platform-independent pretouch memory code.
    // When using THP we need to always pre-touch using small pages as the
    // OS will initially always use small pages.
    return os::vm_page_size();
  }

  const size_t len = pointer_delta(last, first, sizeof(char)) + page_size;
  if (::madvise(first, len, MADV_POPULATE_WRITE) != -1) {
    // Succeeded
    // 0 signals that the platform-independent pretouch memory code should not run.
    return 0;
  }

  // madvise failed
  int err = errno;
  
  log_debug(gc, os)("Called madvise(" PTR_FORMAT ", " SIZE_FORMAT ", %d):"
                    " error='%s' (errno=%d), when THPMode::always=%d and"
                    " UseTransparentHugePages=%d",
                    p2i(first), len, MADV_POPULATE_WRITE, os::strerror(err),
                    err, (int)(HugePages::thp_mode() == THPMode::always),
                    (int)UseTransparentHugePages);

  if (err == EINVAL) {
    // Not supported. When using THP we need to always pre-touch using
    // small pages as the OS will initially always use small pages.
    return os::vm_page_size();
  }

  assert(false, "Unexpected  error='%s' (errno=%d)",os::strerror(err), err);
  return page_size;
}
```

It is unclear to me why a failure with EINVAL should return small pages but other failures should return another page size. Why would we get EINVAL if we already performed the correct checks when setting up UseMadvPopulateWrite? Or is the problem that we skip performing the necessary checks if the user forcefully sets UseMadvPopulateWrite on the command line? Maybe we shouldn't allow that?

Reply:
> It is unclear to me why a failure with EINVAL should return small pages but other failures should return another page size.

Thanks to point it out. It is a mistake.